components/places/src/db/schema.rs

```diff
@@ -220,11 +222,207 @@ const CREATE_IDX_MOZ_BOOKMARKS_PLACELASTMODIFIED: &str = "CREATE INDEX itemlastm
+// origins after places removals. During an INSERT into moz_places, origins are
+// accumulated in this table, then a DELETE FROM moz_updateoriginsinsert_temp
+// will take care of updating the moz_origins table for every new origin. See
+// CREATE_PLACES_AFTERINSERT_TRIGGER_ORIGINS below for details.
+const CREATE_UPDATEORIGINSINSERT_TEMP: &str = "
+    CREATE TEMP TABLE moz_updateoriginsinsert_temp (
+        place_id INTEGER PRIMARY KEY,
+        prefix TEXT NOT NULL,
+        host TEXT NOT NULL,
+        rev_host TEXT NOT NULL,
+        frecency INTEGER NOT NULL
+    )
+";
+
+// This table is used in a similar way to moz_updateoriginsinsert_temp, but for
+// deletes, and triggered via moz_places_afterdelete_trigger.
+//
+// When rows are added to this table, moz_places.origin_id may be null.  That's
+// why this table uses prefix + host as its primary key, not origin_id.
+const CREATE_UPDATEORIGINSDELETE_TEMP: &str = "
+    CREATE TEMP TABLE moz_updateoriginsdelete_temp (
+        prefix TEXT NOT NULL,
+        host TEXT NOT NULL,
+        frecency_delta INTEGER NOT NULL,
+        PRIMARY KEY (prefix, host)
+    ) WITHOUT ROWID
+";
 
+// This table is used in a similar way to moz_updateoriginsinsert_temp, but for
+// updates to places' frecencies, and triggered via
+// moz_places_afterupdate_frecency_trigger.
+//
+// When rows are added to this table, moz_places.origin_id may be null.  That's
+// why this table uses prefix + host as its primary key, not origin_id.
+const CREATE_UPDATEORIGINSUPDATE_TEMP: &str = "
+    CREATE TEMP TABLE moz_updateoriginsupdate_temp (
+        prefix TEXT NOT NULL,
+        host TEXT NOT NULL,
+        frecency_delta INTEGER NOT NULL,
+        PRIMARY KEY (prefix, host)
+    ) WITHOUT ROWID
+";
+
+// Note: desktop places also has a notion of "frecency decay", and it only runs this
+// `WHEN NOT is_frecency_decaying()`.
+const CREATE_PLACES_AFTERUPDATE_FRECENCY_TRIGGER: &str = "
+    CREATE TEMP TRIGGER moz_places_afterupdate_frecency_trigger
+    AFTER UPDATE OF frecency ON moz_places FOR EACH ROW
+    BEGIN
+
+        INSERT INTO moz_updateoriginsupdate_temp (prefix, host, frecency_delta)
+        VALUES (
+            get_prefix(NEW.url),
+            get_host_and_port(NEW.url),
+            MAX(NEW.frecency, 0) - MAX(OLD.frecency, 0)
+        )
+        ON CONFLICT(prefix, host) DO UPDATE
+        SET frecency_delta = frecency_delta + EXCLUDED.frecency_delta;
+    END
+";
+
+// Keys in the moz_meta table. Might be cool to replace this with an enum or something for better
+// type safety but whatever.
+pub(crate) static MOZ_META_KEY_ORIGIN_FRECENCY_COUNT: &'static str = "origin_frecency_count";
+pub(crate) static MOZ_META_KEY_ORIGIN_FRECENCY_SUM: &'static str = "origin_frecency_sum";
+pub(crate) static MOZ_META_KEY_ORIGIN_FRECENCY_SUM_OF_SQUARES: &'static str = "origin_frecency_sum_of_squares";
+
+// This function is a helper for the next several triggers.  It updates the origin
```

Comment:
Two tiny nits:

* Could this be a `///` comment?
* You mentioned an enum for the `moz_meta` keys above; worth using one here for `op`?

Reply:
- Yes, I'll see how I can organize that to make it work
- No, since it's not public and I'd just have to match on it immediately anyway. It's a bit meh, but results in simpler code and it's still internal.